### PR TITLE
[Feature] Port MySQLScanNode to BE connector

### LINKED ISSUE #5735: MySQLScanNode adaptation

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5735

### CODE REVIEW

File: be/src/connector/mysql_connector.cpp [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+    : _scan_node(scan_node), _mysql_scan_node(plan_node.mysql_scan_node) {}
+
+DataSourcePtr MySQLDataSourceProvider::create_data_source(const TScanRange& scan_range) {
+    return std::make_unique<MySQLDataSource>(this, scan_range);
+}
+
+// ================================
+
+MySQLDataSource::MySQLDataSource(const MySQLDataSourceProvider* provider, const TScanRange& scan_range)
+    : _provider(provider), _scan_range(scan_range.) {}
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:31]: There is no mysql scan range.

File: be/src/connector/mysql_connector.cpp [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+Status MySQLDataSource::_init_scanner(RuntimeState *state) {
+    VLOG(1) << "MySQLDataSource::init scanner";
+
+    if (_is_init) {
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:36]: this is called only once.

File: be/src/connector/mysql_connector.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    _tuple_pool = std::make_unique<MemPool>();
+    DCHECK(_tuple_pool != nullptr);
+
+    _is_init = true;
+
+    return Status::OK();
+}
+
+Status MySQLDataSource::open(RuntimeState* state) {
+    VLOG(1) << "MySQLDataSource::Open";
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:69]: consider to remove it.

**Comment by dirtysalt** [comment]
Consider to insert a logical exchange operator at the end of scan operators.

File: be/src/connector/mysql_connector.cpp [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+    return std::make_unique<MySQLDataSource>(this, scan_range);
+}
+
+// ================================
+
+MySQLDataSource::MySQLDataSource(const MySQLDataSourceProvider* provider, const TScanRange& scan_range)
+    : _provider(provider) {}
+
+Status MySQLDataSource::_init_params(RuntimeState *state) {
+    VLOG(1) << "MySQLDataSource::init scanner";
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:34]: Consider refactoring log message.

File: be/src/connector/mysql_connector.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+#include "column/column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "connector/connector.h"
+#include "exec/mysql_scanner.h"
+
+namespace starrocks {
+
+namespace connector {
+#define APPLY_FOR_NUMERICAL_TYPE(M, APPEND_TO_SQL) \
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.h:14]: consider to move this code into cpp file.

File: be/src/connector/mysql_connector.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Tuple id resolved in prepare() to set _tuple_desc;
+    TupleId _tuple_id;
+
+    // select columns
+    std::vector<std::string> _columns;
+    // where clause
+    std::vector<std::string> _filters;
+    // limit for query with external table.
+    int64_t _limit;
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.h:89]: this variable is stored in `DataSourc` class:
1. conjunct
2. runtime filters
3. read limit

File: be/src/connector/mysql_connector.cpp [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+Status MySQLDataSource::_init_params(RuntimeState *state) {
+    VLOG(1) << "MySQLDataSource::init mysql scan params";
+
+    DCHECK(state != nullptr);
+
+    _columns = _provider->_mysql_scan_node.columns;
+    _filters = _provider->_mysql_scan_node.filters;
+
+    _table_name = _provider->_mysql_scan_node.table_name;
+    _read_limit = -1;
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:55]: no need to set it. This variable has been set by framework.

File: be/src/connector/mysql_connector.cpp [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+
+Status MySQLDataSource::get_next(RuntimeState* state, vectorized::ChunkPtr* chunk) {
+    VLOG(1) << "MySQLDataSource::GetNext";
+
+    DCHECK(state != nullptr && chunk != nullptr);
+
+    RETURN_IF_CANCELLED(state);
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+
+    // TODO: check limit
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:209]: No need to check limit. It's been handled by frame work.

File: be/src/connector/mysql_connector.cpp [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+    if (_is_finished) {
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:210]: No need to maintain this variable.  Return EOF, then get_next will not be called any more.

File: be/src/connector/mysql_connector.cpp [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+        return Status::EndOfFile("");
+    }
+
+    vectorized::Chunk* ck = chunk->get();
+    std::vector<SlotDescriptor*> slot_descs = _tuple_desc->slots();
+    // init column information
+    for (auto& slot_desc : slot_descs) {
+        ColumnPtr column = ColumnHelper::create_column(slot_desc->type(), slot_desc->is_nullable());
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:218]: Check `ChunkHelper::create_chunk`.

File: be/src/connector/mysql_connector.cpp [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+        size_t* length = nullptr;
+        RETURN_IF_ERROR(_mysql_scanner->get_next_row(&data, &length, &mysql_eos));
+        if (mysql_eos) {
+            _is_finished = true;
+            return Status::EndOfFile("");
+        }
+
+        ++row_num;
+
+        int materialized_col_idx = -1;
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:244]: Could following logic/code be put into a function, like `build_chunk` ?

File: be/src/connector/mysql_connector.cpp [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+    _my_param.host = mysql_table->host();
+    _my_param.port = mysql_table->port();
+    _my_param.user = mysql_table->user();
+    _my_param.passwd = mysql_table->passwd();
+    _my_param.db = mysql_table->mysql_db();
+    // new one scanner
+    _mysql_scanner = std::make_unique<MysqlScanner>(_my_param);
+    DCHECK(_mysql_scanner != nullptr);
+
+    _tuple_pool = std::make_unique<MemPool>();
```
Comment (dirtysalt) [inline be/src/connector/mysql_connector.cpp:78]: not used any more.

**Comment by dirtysalt** [comment]
LGTM. I'll run a pr regression for you.

**Review by dirtysalt — DISMISSED** [review]

File: be/src/connector/mysql_connector.cpp [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    bool parse_success = true;
+
+    // Parse the raw-text data. Translate the text string to internal format.
+    switch (slot_desc->type().type) {
+    case TYPE_VARCHAR:
+    case TYPE_CHAR: {
+        Slice value(data, len);
+        dynamic_cast<BinaryColumn*>(data_column)->append(value);
```
Comment (imay) [inline be/src/connector/mysql_connector.cpp:298]: dynamic_cast will cause bad performance if you call it for each column. Maybe you can use represent_cast

File: be/src/connector/mysql_connector.cpp [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+
+    std::unordered_map<std::string, std::vector<std::string>> filters_in;
+    std::unordered_map<std::string, bool> filters_null_in_set;
+
+    // In Filter have been put into _conjunct_ctxs,
+    // so we iterate all ExprContext to use it.
+    for (int i = 0; i < _conjunct_ctxs.size(); ++i) {
+        ExprContext* ctx = _conjunct_ctxs[i];
+        const Expr* root_expr = ctx->root();
+        if (root_expr != nullptr) {
```
Comment (imay) [inline be/src/connector/mysql_connector.cpp:99]:
```suggestion
        if (root_expr == nullptr) {
            continue;
        }
     
```
make code line shorter

**Review by imay — DISMISSED** [review]

**Comment by dirtysalt** [comment]
pr regression looks ok. Myabe you can put another commit to remove mysql scan node/operator.

**Review by dirtysalt — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**